src/unitxt/metrics.py

```diff
@@ -366,16 +437,201 @@ def compute(
+        assert (
+            "agg_func" in fields
+        ), "fields should have a key 'agg_func' whose value is a 3-element list of a function name, function definition, and a boolean indicator"
+        assert isinstance(
+            fields["agg_func"], list
+        ), "fields['agg_func'] should be a list"
+        assert (
+            len(fields["agg_func"]) == 3
+        ), "fields['agg_func'] should be a 3-element list"
+        assert isinstance(
+            fields["agg_func"][0], str
+        ), "first item in fields['agg_func'] should be a string name of a function"
+        assert callable(
+            fields["agg_func"][1]
+        ), "second item in fields['agg_func'] should be a callable function"
+        assert isinstance(
+            fields["agg_func"][2], bool
+        ), "third item in fields['agg_func'] should be a boolean value"
+        if "score_fields" in fields:
+            assert isinstance(fields["score_fields"], list)
+
+        # for aggregations that conduct a comparison, expects a 'variant_type' field
+        if self._function_uses_variant_types(fields["agg_func"][1]):
+            assert all(
+                "variant_type" in instance["additional_inputs"]
+                for instance in instances
+            ), f"since group_mean aggregation function {fields['agg_func'][1]} performs a comparison, each instance's additional_inputs dict must have a key variant_type"
+
     def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generator:
+        instances, global_score = self.compute_instance_scores(stream)
+        from copy import deepcopy
+
+        for reduction_type, reduction_params in self.reduction_map.items():
+            assert (
+                reduction_type in self.implemented_reductions
+            ), f"Reduction {reduction_type} is not implemented, use one of {self.implemented_reductions}"
+
+            field_name_full_prefix = ""
+            # used for passing to the bootstrapping, depends on whether the groups are fixed or not
+            aggregation_function = self.average_item_scores
+            if reduction_type == "mean":
+                reduction_fields = list(set(reduction_params))
+                # extract only the dict of instance scores
+                scores_to_resample = deepcopy(instances)
+            elif reduction_type == "group_mean":
+                self._validate_group_mean_reduction(instances=instances)
+                reduction_fields = (
+                    [self.main_score]
+                    if "score_fields" not in reduction_params
+                    else list(set(reduction_params["score_fields"]))
+                )
+                aggregation_function_name = str(reduction_params["agg_func"][0])
+                field_name_full_prefix = "group_" + aggregation_function_name + "_"
+                if reduction_params["agg_func"][2]:
+                    field_name_full_prefix = "fixed_" + field_name_full_prefix
```

Comment:
because it indicates the the groups are resampled as fixed units (when reduction_params["agg_func"][2] is True), rather than resampled as instances and then grouped.  I wanted to differentiate the two.

Reply:
i see, sounds good. maybe just a small suggestion, add another word like `fixed_groups_` or `sampled_groups_`. If you agree it improves readability.